File: crates/nmp-cli/registry/swiftui/content-kind-registry/EmbedKindProjection.swift
import Foundation

/// Swift Codable mirror of the Rust `nmp_content::embed_projection::EmbedKindProjection`.
///
/// The Rust enum is serialized with `#[serde(tag = "variant", content = "data",
/// rename_all = "camelCase")]`, so JSON looks like:
///   { "variant": "article", "data": { "id": "…", "title": "…", … } }
///
/// Each variant is a typed payload the native renderer consumes verbatim; the
/// dispatch decision is `match projection` on the Swift side.
public enum EmbedKindProjection: Equatable {
    case shortNote(ShortNoteProjection)
    case article(ArticleProjection)
    case highlight(HighlightProjection)
    case profile(ProfileProjection)
    case unknown(UnknownProjection)
}

/// kind:1 short text note projection.
public struct ShortNoteProjection: Equatable {
    public let id: String
    public let authorPubkey: String
    public let authorDisplayName: String?
    public let authorPictureUrl: String?
    public let createdAt: UInt64
    /// Plain-text fallback for the content body. The Swift mirror does NOT
    /// re-implement the Rust content tokenizer; renderers read `content` as a
    /// plain string and let SwiftUI handle the layout.
    public let content: String
    public let mediaUrls: [String]

    public init(
        id: String,
        authorPubkey: String,
        authorDisplayName: String? = nil,
        authorPictureUrl: String? = nil,
        createdAt: UInt64 = 0,
        content: String = "",
        mediaUrls: [String] = []
    ) {
        self.id = id
        self.authorPubkey = authorPubkey
        self.authorDisplayName = authorDisplayName
        self.authorPictureUrl = authorPictureUrl
        self.createdAt = createdAt
        self.content = content
        self.mediaUrls = mediaUrls
    }
}

/// kind:30023 long-form article projection (NIP-23).
public struct ArticleProjection: Equatable {
    public let id: String
    public let authorPubkey: String
    public let authorDisplayName: String?
    public let authorPictureUrl: String?
    public let createdAt: UInt64
    public let title: String?
    public let summary: String?
    public let heroImageUrl: String?
    public let dTag: String
    /// Plain-text body fallback (the Rust resolver also emits a content tree;
    /// the Swift gallery showcase renders title/summary/hero image only).
    public let content: String

    public init(
        id: String,
        authorPubkey: String,
        authorDisplayName: String? = nil,
        authorPictureUrl: String? = nil,
        createdAt: UInt64 = 0,
        title: String? = nil,
        summary: String? = nil,
        heroImageUrl: String? = nil,
        dTag: String = "",
        content: String = ""
    ) {
        self.id = id
        self.authorPubkey = authorPubkey
        self.authorDisplayName = authorDisplayName
        self.authorPictureUrl = authorPictureUrl
        self.createdAt = createdAt
        self.title = title
        self.summary = summary
        self.heroImageUrl = heroImageUrl
        self.dTag = dTag
        self.content = content
    }
}

/// kind:9802 highlight projection (NIP-84).
public struct HighlightProjection: Equatable {
    public let id: String
    public let authorPubkey: String
    public let authorDisplayName: String?
    public let createdAt: UInt64
    public let highlightedText: String
    public let sourceEventId: String?
    public let sourceEventAddr: String?
    public let sourceUrl: String?
    public let context: String?

    public init(
        id: String,
        authorPubkey: String,
        authorDisplayName: String? = nil,
        createdAt: UInt64 = 0,
        highlightedText: String = "",
        sourceEventId: String? = nil,
        sourceEventAddr: String? = nil,
        sourceUrl: String? = nil,
        context: String? = nil
    ) {
        self.id = id
        self.authorPubkey = authorPubkey
        self.authorDisplayName = authorDisplayName
        self.createdAt = createdAt
        self.highlightedText = highlightedText
        self.sourceEventId = sourceEventId
        self.sourceEventAddr = sourceEventAddr
        self.sourceUrl = sourceUrl
        self.context = context
    }
}

/// kind:0 profile metadata projection.
public struct ProfileProjection: Equatable {
    public let pubkey: String
    public let displayName: String?
    public let pictureUrl: String?
    public let about: String?
    public let nip05: String?
    public let lud16: String?
    public let bannerUrl: String?

    public init(
        pubkey: String,
        displayName: String? = nil,
        pictureUrl: String? = nil,
        about: String? = nil,
        nip05: String? = nil,
        lud16: String? = nil,
        bannerUrl: String? = nil
    ) {
        self.pubkey = pubkey
        self.displayName = displayName
        self.pictureUrl = pictureUrl
        self.about = about
        self.nip05 = nip05
        self.lud16 = lud16
        self.bannerUrl = bannerUrl
    }
}

/// Fallback projection for kinds without a registered handler.
public struct UnknownProjection: Equatable {
    public let kind: UInt32
    public let authorPubkey: String
    public let authorDisplayName: String?
    public let authorPictureUrl: String?
    public let createdAt: UInt64
    public let content: String
    public let tags: [[String]]
    public let altText: String?

    public init(
        kind: UInt32,
        authorPubkey: String,
        authorDisplayName: String? = nil,
        authorPictureUrl: String? = nil,
        createdAt: UInt64 = 0,
        content: String = "",
        tags: [[String]] = [],
        altText: String? = nil
    ) {
        self.kind = kind
        self.authorPubkey = authorPubkey
        self.authorDisplayName = authorDisplayName
        self.authorPictureUrl = authorPictureUrl
        self.createdAt = createdAt
        self.content = content
        self.tags = tags
        self.altText = altText
    }
}

/// Full envelope mirror of `nmp_content::embed_projection::EmbeddedEventEnvelope`.
public struct EmbeddedEventEnvelope: Equatable {
    /// The original nostr: URI (nevent1… / naddr1… / npub1…).
    public let uri: String
    /// Primary identifier: event-id hex for event-addressed refs, or
    /// `"kind:pubkey:d"` coordinate string for addressable events.
    public let primaryId: String
    /// Recursion guard state. Currently surfaced as defaults; reserved for
    /// nested-embed depth limits.
    public let depth: UInt8
    public let maxDepth: UInt8
    /// Kind-dispatched projection — drives which native renderer is chosen.
    public let projection: EmbedKindProjection
    /// Whether this embed should be collapsed (depth limit, cycle, unsupported).
    public let collapsed: Bool
    /// Optional machine-readable collapse reason.
    public let collapseReason: String?

    public init(
        uri: String,
        primaryId: String,
        depth: UInt8 = 0,
        maxDepth: UInt8 = 4,
        projection: EmbedKindProjection,
        collapsed: Bool = false,
        collapseReason: String? = nil
    ) {
        self.uri = uri
        self.primaryId = primaryId
        self.depth = depth
        self.maxDepth = maxDepth
        self.projection = projection
        self.collapsed = collapsed
        self.collapseReason = collapseReason
    }
}

File: apps/nmp-gallery/android/app/src/main/kotlin/org/nmp/gallery/registry/EmbedKindProjection.kt
// Requires: kotlinx.serialization. Kotlin 1.9+.
//
// Kotlin kotlinx.serialization mirror of the Rust
// `nmp_content::embed_projection::EmbeddedEventEnvelope` / `EmbedKindProjection`.
//
// The Rust enum is serialized with `#[serde(tag = "variant", content = "data",
// rename_all = "camelCase")]`, but the typed NEMB sidecar decoder on Android
// populates exactly one variant field (`shortNote` / `article` / …). This file
// models that already-decoded shape: the dispatch decision is a `when` over
// which variant is non-null. No protocol parsing happens here — the kernel
// resolved the event into a typed projection before it reached this layer.
//
// Compose mirror of the SwiftUI `EmbedKindProjection.swift`.

package org.nmp.gallery.registry

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Kind-dispatched embed projection — exactly one variant is non-null, selected
 * by the kernel's resolver from the `kind` discriminant. Mirrors the SwiftUI
 * `EmbedKindProjection` enum. DECODE-ONLY: no resolution logic lives here.
 */
@Serializable
public data class EmbedKindProjection(
    @SerialName("short_note") val shortNote: ShortNoteProjection? = null,
    val article: ArticleProjection? = null,
    val highlight: HighlightProjection? = null,
    val profile: ProfileProjection? = null,
    val unknown: UnknownProjection? = null,
)

/** kind:1 short text note projection. */
@Serializable
public data class ShortNoteProjection(
    val id: String = "",
    @SerialName("author_pubkey") val authorPubkey: String = "",
    @SerialName("author_display_name") val authorDisplayName: String? = null,
    @SerialName("author_picture_url") val authorPictureUrl: String? = null,
    @SerialName("created_at") val createdAt: Long = 0,
    /** Plain-text fallback for the content body. */
    val content: String = "",
    @SerialName("media_urls") val mediaUrls: List<String> = emptyList(),
)

/** kind:30023 long-form article projection (NIP-23). */
@Serializable
public data class ArticleProjection(
    val id: String = "",
    @SerialName("author_pubkey") val authorPubkey: String = "",
    @SerialName("author_display_name") val authorDisplayName: String? = null,
    @SerialName("author_picture_url") val authorPictureUrl: String? = null,
    @SerialName("created_at") val createdAt: Long = 0,
    val title: String? = null,
    val summary: String? = null,
    @SerialName("hero_image_url") val heroImageUrl: String? = null,
    @SerialName("d_tag") val dTag: String = "",
    val content: String = "",
)

/** kind:9802 highlight projection (NIP-84). */
@Serializable
public data class HighlightProjection(
    val id: String = "",
    @SerialName("author_pubkey") val authorPubkey: String = "",
    @SerialName("author_display_name") val authorDisplayName: String? = null,
    @SerialName("created_at") val createdAt: Long = 0,
    @SerialName("highlighted_text") val highlightedText: String = "",
    @SerialName("source_event_id") val sourceEventId: String? = null,
    @SerialName("source_event_addr") val sourceEventAddr: String? = null,
    @SerialName("source_url") val sourceUrl: String? = null,
    val context: String? = null,
)

/** kind:0 profile metadata projection. */
@Serializable
public data class ProfileProjection(
    val pubkey: String = "",
    @SerialName("display_name") val displayName: String? = null,
    @SerialName("picture_url") val pictureUrl: String? = null,
    val about: String? = null,
    val nip05: String? = null,
    val lud16: String? = null,
    @SerialName("banner_url") val bannerUrl: String? = null,
)

/** Fallback projection for kinds without a registered handler. */
@Serializable
public data class UnknownProjection(
    val kind: Int = 0,
    @SerialName("author_pubkey") val authorPubkey: String = "",
    @SerialName("author_display_name") val authorDisplayName: String? = null,
    @SerialName("author_picture_url") val authorPictureUrl: String? = null,
    @SerialName("created_at") val createdAt: Long = 0,
    val content: String = "",
    val tags: List<List<String>> = emptyList(),
    @SerialName("alt_text") val altText: String? = null,
)

/**
 * Full envelope mirror of `nmp_content::embed_projection::EmbeddedEventEnvelope`.
 * Populated exclusively from the typed NEMB sidecar (not a JSON projection).
 */
@Serializable
public data class EmbeddedEventEnvelope(
    /** The original nostr: URI (nevent1… / naddr1… / npub1…). */
    val uri: String = "",
    /** Primary identifier: event-id hex, or `kind:pubkey:d` coordinate. */
    @SerialName("primary_id") val primaryId: String = "",
    /** Recursion guard state (reserved for nested-embed depth limits). */
    val depth: Int = 0,
    @SerialName("max_depth") val maxDepth: Int = 4,
    /** Kind-dispatched projection — drives which renderer is chosen. */
    val projection: EmbedKindProjection? = null,
    /** Whether this embed should be collapsed (depth limit, cycle, unsupported). */
    val collapsed: Boolean = false,
    /** Optional machine-readable collapse reason. */
    @SerialName("collapse_reason") val collapseReason: String? = null,
)
